curl/inc/leatherman/curl/client.hpp

```diff
@@ -0,0 +1,217 @@
+
+     private:
+        static void cleanup(char const* str);
+    };
+
+    /**
+     * The exception for HTTP.
+     */
+    struct http_exception : std::runtime_error
+    {
+        /**
+         * Constructs an http_exception.
+         * @param message The exception message.
+         */
+        http_exception(std::string const &message) :
+            runtime_error(message)
+        {
+        }
+    };
+
+    /**
+     * The exception for HTTP requests.
+     */
+    struct http_request_exception : http_exception
+    {
+        /**
+         * Constructs an http_request_exception.
+         * @param req The HTTP request that caused the exception.
+         * @param message The exception message.
+         */
+        http_request_exception(request req, std::string const &message) :
+            http_exception(message),
+            _req(std::move(req))
+        {
+        }
+
+        /**
+         * Gets the request associated with the exception
+         * @return Returns the request associated with the exception.
+         */
+        request const& req() const
+        {
+            return _req;
+        }
+
+     private:
+        request _req;
+    };
+
+    /**
+     * Implements a client for HTTP.
+     * Note: this class is not thread-safe.
+     */
+    struct client
+    {
+        /**
+         * Constructs an HTTP client.
+         */
+        client();
+
+        /**
+         * Moves the given client into this client.
+         * @param other The client to move into this client.
+         */
+        client(client&& other);
+
+        /**
+         * Moves the given client into this client.
+         * @param other The client to move into this client.
+         * @return Returns this client.
+         */
+        client& operator=(client&& other);
+
+        /**
+         * Performs a GET with the given request.
+         * @param req The HTTP request to perform.
+         * @return Returns the HTTP response.
+         */
+        response get(request const& req);
+
+        /**
+         * Performs a POST with the given request.
+         * @param req The HTTP request to perform.
+         * @return Returns the HTTP response.
+         */
+        response post(request const& req);
+
+        /**
+         * Performs a PUT with the given request.
+         * @param req The HTTP request to perform.
+         * @return Returns the HTTP response.
+         */
+        response put(request const& req);
+
+        /**
+         * Sets the path to the CA certificate file.
+         * @param cert_file The path to the CA certificate file.
+         */
+        void set_ca_cert(std::string cert_file);
+
+        /**
+         * Set client SSL certificate and key.
+         * @param client_cert The path to the client's certificate file.
+         * @param client_key The path to the client's key file.
+         */
+        void set_client_cert(std::string client_cert, std::string client_key);
```

Comment:
This member should take the arguments by `const&` since it isn't moving either argument in its implementation.